### PR TITLE
Feat/spec object wrapper

### PR DESCRIPTION
**Pull request by travisliu**
Why need this change ? / Root Cause :
--------------
The validate_data_spec method is great. But encapsulating spec object can makes that easier to use.

Changed Made :
--------------
This pull request only simply add an object wrapper into wrappers.py, and import it in __init__.py

- data_spec_validator/spec/wrappers.py
- data_spec_validator/spec/__init__.py

Test Scope / Change Impact :
--------------
There are only new wrappers added without any change of original codebase. So  no change Impact should be happened.

Any comment or suggestion is welcome

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -141,6 +141,22 @@ validate_data_spec(ok_data, GreaterThanSpec) # raise Exception
 "field: XXX not well-formatted"
 ```
 
+### Spec Object Wrapper
+
+- Inherit Spec wrapper, the spec object can simply valdates data with a class method. (see [test_wrapped_spec_class_validate](https://github.com/travisliu/data-spec-validator/blob/feat/spec-object-wrapper/test/test_spec.py#L993) for more)
```
Comment (kilikkuo) [inline README.md:146]: s/valdates/validate/g

File: README.md [hunk 2]
```diff
@@ -141,6 +141,22 @@ validate_data_spec(ok_data, GreaterThanSpec) # raise Exception
...
+from data_spec_validator.spec import (
+    INT,
+    Spec,
+)
+
+class SomeSpec(Spec):
+    field_a = Checker([INT])
+
+some_data = dict(field_a=4)
+SomeSpec.validate(some_data) # return True
```
Comment (kilikkuo) [inline README.md:157]: Nice!

File: README.md (hunk 1 above)
Comment (travisliu) [reply README.md]: that's embarrassing. The typo has been corrected!

File: data_spec_validator/spec/wrappers.py [hunk 3]
```diff
@@ -14,3 +14,32 @@ def validate(self, value, extra, data):
...
+            setattr(spec, attribute, getattr(cls, attribute))
+
+        return spec
+
+    @classmethod
+    def validate(cls, data, **kwargs):
+        if (cls.__spec_object is None):
+            cls.__spec_object = cls.__create_specObject()
+
+        return validate_data_spec(data, cls.__spec_object, **kwargs)
```
Comment (xeonchen) [inline data_spec_validator/spec/wrappers.py:44]: Could you please briefly describe the reason why the `SpecObject` type and its instance is required here?

Comment (travisliu) [reply data_spec_validator/spec/wrappers.py:44]: That's a good question. Because I added a new validate function. If I inject that cls variable into the validate_data_spec in #44 without new spec instance.  "validate" will be return as a field that needs to be validated by  [_extract_fields](https://github.com/hardcoretech/data-spec-validator/blob/7ee0944ca9899d565ad04ed82ca26bb402970958/data_spec_validator/spec/validators.py#L47), which will cause an error. 

In the first place, I was trying to make some small changes for this repo. So I work around to avoiding modify that _extract_fields function. Overriding` __dir__` is another way has been tried. But it didn't change any result  from `dir(Spec)`, only works for `dir(Spec())`.

In summary, that's why I created a new spec class for validate_data_spec.  The new  instance is because we don't want side effects from other spec object. That can keep it isolated. This way might not be a best solution, any suggestion is welcome! 😃

Comment (travisliu) [reply data_spec_validator/spec/wrappers.py:44]: BTW, don't feel bad if you guys want to reject this pull request. this pr more like a  side project. It let me can give it a try to Python.

Comment (kilikkuo) [reply data_spec_validator/spec/wrappers.py:44]: I like the idea `SomeSpec.validate(data, **kwargs)`.

Though there are two concerns to me, the first one is the derived class(and its parents') members (i.e. attributes, properties...) are also listed when a call `dir()` on it.  That's the reason why `def __create_specObject` is here.
The other one is I don't want to create any Spec instance as it should be just a descriptor-like class to indicate what fields to validate.

Instead of calling `dir()`,  it seems valuable enhancing the `_extract_fields` by using `SOME_CLASS.__dict__.keys()` to keep leaf-class-level attributes (parents' attributes are excluded).  
So that the `class Spec` may just have one method `validate`
```
    @classmethod
    def validate(cls, data, **kwargs):
        return validate_data_spec(data, cls, **kwargs)
```

Comment (xeonchen) [reply data_spec_validator/spec/wrappers.py:44]: OK, what you do here is to make `_extract_fields` work as usual by passing in a subclass without listing fields of the base class `Spec`, right?

I'd suggest to modify `_extract_fields` and make it either
* support both classes: general classes and `Spec` subclasses, or
* check if the class has some specific method, such as `list_fields`.

This will make the code more readable, and you don't need those `setattr`s or any instance at all.

Comment (travisliu) [reply data_spec_validator/spec/wrappers.py]: hi @kilikkuo and @xeonchen,

Thanks for your suggestions. I figured out the metaclass usage and use it to organize a spec_field. that can be used in _extract_fields function.

About changes in _extract_fields, I thought the condition should be `isinstance(checker, Spec)`, but I can import Spec into validators.py. So I used `hasattr(checker, 'spec_fields')` instead. it could be a potential bug if other general class also has the same name field need to be validated. so I named the spec_fields field as uncommon as possible.

that's pretty much, any comment or suggestions is still welcome.

Comment (travisliu) [reply data_spec_validator/spec/wrappers.py]: There is another way, all we have to do just only replace one line of code like:
```python
def _extract_fields(checker):
+    return list(filter(lambda f: isinstance(getattr(checker, f), Checker), dir(checker)))
-    return list(filter(lambda f: type(f) == str and not (f.startswith('__') and f.endswith('__')), dir(checker)))
```

that's the simplest approach. we even don't need the SpecMeta metaclass, but it could changed origin behavior. I not sure how you used it in production. this change could by risky. That's the reason why I tired to walk around.
